[rados] Development of RADOS binding of Ceph FS

jaemyoun opened the pull request:
Development of RADOS binding of Ceph File System
Even though Ceph FS is not NoSQL database, RADOS, Ceph utility, supports object-based storage.

For Ceph benchmark for cloud services. I implemented RADOS binding using [rados-java library](https://github.com/ceph/rados-java) expect the scan() function.

busbey commented on pom.xml:
```diff
@@ -93,6 +93,8 @@ LICENSE file.
     <aerospike.version>3.1.2</aerospike.version>
     <solr.version>5.4.0</solr.version>
+    <rados.version>0.3.0-SNAPSHOT</rados.version>
+    <json.version>20160212</json.version>
```
this should just be set in the pom for the rados client, since it needn't be shared across different datastore bindings..

busbey commented on rados/README.md:
```diff
@@ -0,0 +1,64 @@
+<!--
+Copyright (c) 2014 - 2015 YCSB contributors. All rights reserved.
```
years need to be corrected.

busbey commented on rados/README.md:
```diff
@@ -0,0 +1,64 @@
+
+Git clone YCSB and compile:
+
+    git clone -b dev-ceph-rados https://github.com/jaemyoun/YCSB.git
```
should probably just tell folks to download. Definitely should refer to the correct git repo if it stays as a clone.

busbey commented on rados/pom.xml:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
needs a file header with licensing info

busbey commented on rados/src/main/java/com/yahoo/ycsb/db/RadosClient.java:
```diff
@@ -0,0 +1,146 @@
+package com.yahoo.ycsb.db;
```
needs a file header with licensing info

jaemyoun replied:
fixed it

busbey commented:
The travis failure is because you need to add the new module to the list of dependencies for the distribution module.

jaemyoun commented:
I fixed all code you mentioned, and an additional error.

busbey commented on rados/pom.xml:
```diff
@@ -0,0 +1,56 @@
+  </dependencies>
+
+  <properties>
+  	<rados.version>0.3.0-SNAPSHOT</rados.version>
```
We don't rely on SNAPSHOT versions in our releases. Can we rely on a prior version? If not, when is 0.3.0 expected to come out?

jaemyoun commented:
I changed the version to 0.2.0 and merged from YCSB master for updating to 0.9.0-SNAPSHOT.

busbey commented on rados/src/main/java/com/yahoo/ycsb/db/RadosClient.java:
```diff
@@ -0,0 +1,167 @@
+      pool = POOL_DEFAULT;
+    }
+
+    rados = new Rados(id);
```
This needs to be protected by a lock, sine rados is static and we instantiate a client for each thread.

Presuming the rados and ioctx objects are themselves threadsafe, then you should synchronize and use a threadcount to decide when it's safe to do shutdown in the `cleanup` method.

jaemyoun replied:
Did you mean these to change Static Strings? The three value will be different depend on user setting.

busbey replied:
No, I mean the member variables `rados` and `ioctx` are static. We instantiate a RadosClient per worker thread in the client, so access to initializing and cleaning up these static members must be protected by a lock.

If the member classes themselves aren't threadsafe, then you'll need to either guard all uses with a lock or switch to making them non-static.

busbey commented:
does rados provide any kind of testing utility jar that could be used to include a test?

jaemyoun commented:
I switched `rados` and `ioctx`to non-static, and implemented a test.
The test is success in my cluster, but the Travis CI building was failed because there is no the Ceph cluster.

busbey commented:
> The test is success in my cluster, but the Travis CI building was failed because there is no the Ceph cluster.

Does Ceph have some kind of artifact for making a local test instance that we could make use of? (see the `hbase10` or `accumulo` bindings for examples) If not,  we should update the test to make use of the JUnit `Assume` API to skip the test when a Ceph instance is not present (see the `mongodb` or `riak` bindings as an example).

jaemyoun commented:
> If not, we should update the test to make use of the JUnit Assume API to skip the test when a Ceph instance is not present (see the mongodb or riak bindings as an example).

I was not able to find the test instance for Ceph, so I updated to skip the test using `JUnit AssumeNoException`.
